### PR TITLE
[Datasets] Support tensor columns in `to_tf` and `to_torch`.

### LINKED ISSUE #18314: [Datasets] Find workaround for ingesting tensor columns into pytorch.
Ingesting tables with tensor columns into pytorch via `ds.to_torch()` is blocked by pytorch supporting tensor creation from objects that implement the `__array__` interface. See [issue](https://github.com/pytorch/pytorch/issues/51156).

## Possible Workarounds

1. Manually check in `ds.to_torch()` for extension array columns and manually convert it to an ndarray via invoking the `__array__` protocol.

### PR DESCRIPTION
**Pull request by clarkzinzow**
This PR adds support for tensor columns in the `to_tf()` and `to_torch()` APIs.

For Torch, this involves an explicit extension array check and (zero-copy) conversion of the tensor column to a NumPy array before converting the column to a Torch tensor.

For TensorFlow, this involves bypassing `df.values` when converting tensor feature columns to NumPy arrays, instead manually creating a single NumPy array from the column Series.

In both cases, I think that the UX around heterogeneous feature columns and squeezing the column dimension could be improved, but I'm saving that for a future PR.

## Related issue number

Closes https://github.com/ray-project/ray/issues/18314, https://github.com/ray-project/ray/issues/18315

## Checks

- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [x] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [x] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [x] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: python/ray/data/dataset.py [hunk 1]
```diff
@@ -2380,6 +2380,27 @@ def to_tf(
         if isinstance(output_signature, list):
             output_signature = tuple(output_signature)
 
+        def get_df_values(df: "pandas.DataFrame") -> np.ndarray:
+            # TODO(Clark): Support unsqueezing column dimension API, simialr to
+            # to_torch().
+            try:
```
Comment (amogkam) [inline python/ray/data/dataset.py:2386]: Could you add a comment explaining what this try-except is checking for?

File: python/ray/data/tests/test_dataset.py [hunk 2]
```diff
@@ -950,49 +950,83 @@ def test_tensors_in_tables_parquet_bytes_with_schema(
...
+            "label": [4.0, 5.0, 6.0],
+        }
     )
     df = pd.concat([df1, df2])
     ds = ray.data.from_pandas([df1, df2])
     ds = maybe_pipeline(ds, pipelined)
     torchd = ds.to_torch(label_column="label", batch_size=2)
 
-    num_epochs = 2
+    num_epochs = 1 if pipelined else 2
```
Comment (amogkam) [inline python/ray/data/tests/test_dataset.py:976]: Does pipelined case not work with 2 epochs?

File: python/ray/data/tests/test_dataset.py [hunk 3]
```diff
@@ -950,49 +950,83 @@ def test_tensors_in_tables_parquet_bytes_with_schema(
...
     for _ in range(num_epochs):
         iterations = []
         for batch in iter(torchd):
-            iterations.append(torch.cat((*batch[0], batch[1]), axis=1).numpy())
```
Comment (amogkam) [inline python/ray/data/tests/test_dataset.py:985]: do we not want to check label column anymore?

**Review by amogkam — COMMENTED** [review]
Nice! Left some comments, mostly for my understanding!

File: doc/source/data/dataset-tensor-support.rst [hunk 4]
```diff
@@ -246,5 +246,3 @@ This feature currently comes with a few known limitations that we are either act
 
  * All tensors in a tensor column currently must be the same shape. Please let us know if you require heterogeneous tensor shape for your tensor column! Tracking issue is `here <https://github.com/ray-project/ray/issues/18316>`__.
  * Automatic casting via specifying an override Arrow schema when reading Parquet is blocked by Arrow supporting custom ExtensionType casting kernels. See `issue <https://issues.apache.org/jira/browse/ARROW-5890>`__. An explicit ``tensor_column_schema`` parameter has been added for :func:`read_parquet() <ray.data.read_api.read_parquet>` as a stopgap solution.
- * Ingesting tables with tensor columns into pytorch via ``ds.to_torch()`` is blocked by pytorch supporting tensor creation from objects that implement the `__array__` interface. See `issue <https://github.com/pytorch/pytorch/issues/51156>`__. Workarounds are being `investigated <https://github.com/ray-project/ray/issues/18314>`__.
```
Comment (ericl) [inline doc/source/data/dataset-tensor-support.rst:249]: :+1: doc updates. Could we also add some explicit examples / text in the docs on how to do this?

Comment (clarkzinzow) [reply doc/source/data/dataset-tensor-support.rst:249]: Hmm document how to do what? This should now transparently work with tensor columns without any considerations by the user.

File: python/ray/data/dataset.py [hunk 5]
```diff
@@ -2380,6 +2380,27 @@ def to_tf(
         if isinstance(output_signature, list):
             output_signature = tuple(output_signature)
 
+        def get_df_values(df: "pandas.DataFrame") -> np.ndarray:
+            # TODO(Clark): Support unsqueezing column dimension API, simialr to
```
[the remark this answers is not in the document]
Comment (clarkzinzow) [reply python/ray/data/dataset.py:2384]: Oof thanks!

File: python/ray/data/dataset.py (hunk 1 above)
Comment (clarkzinzow) [reply python/ray/data/dataset.py:2386]: There's a comment on the need for this exception handling in the except clause below.

Comment (amogkam) [reply python/ray/data/dataset.py:2386]: Oh oops, that sounds good!

File: python/ray/data/tests/test_dataset.py (hunk 2 above)
Comment (clarkzinzow) [reply python/ray/data/tests/test_dataset.py:976]: You can't consume a pipeline multiple times, this test never worked with `num_epochs = 2` and pipelining. See `test_to_torch` for this same pattern.

File: python/ray/data/tests/test_dataset.py (hunk 3 above)
Comment (clarkzinzow) [reply python/ray/data/tests/test_dataset.py:985]: There isn't really a need to check the label column since it's orthogonal to this test, but I can add it back.

File: python/ray/data/dataset.py [hunk 6]
```diff
@@ -2380,6 +2380,27 @@ def to_tf(
...
+
+                # Pandas DataFrame.values doesn't support extension arrays in all
+                # supported Pandas versions, so we check to see if this DataFrame
+                # contains any extensions arrays and do a manual conversion if so.
+                # See https://github.com/pandas-dev/pandas/pull/43160.
+                if any(
+                    isinstance(dtype, pd.api.extensions.ExtensionDtype)
+                    for dtype in df.dtypes
+                ):
+                    values = np.stack([col.to_numpy() for _, col in df.items()], axis=1)
```
Comment (amogkam) [inline python/ray/data/dataset.py:2399]: hmm do we want to conver all columns to numpy if _any_ column is an ExtensionDtype.

Or should we only do col.to_numpy() only if the dtype of that particular columns is an ExtensionDtype

Comment (clarkzinzow) [reply python/ray/data/dataset.py:2399]: It doesn't matter, `df.values` converts all columns to NumPy ndarrays anyway so it would be exactly the same.

File: python/ray/data/tests/test_dataset.py (hunk 3 above)
Comment (amogkam) [reply python/ray/data/tests/test_dataset.py:985]: got it ok, that sounds good!

Comment (clarkzinzow) [reply python/ray/data/tests/test_dataset.py:985]: Added the labels check back!

Comment (clarkzinzow) [reply python/ray/data/tests/test_dataset.py:985]: Ah I only added it back for `to_tf`, let me add it back for `to_torch` too.

File: doc/source/data/dataset-tensor-support.rst (hunk 4 above)
Comment (ericl) [reply doc/source/data/dataset-tensor-support.rst:249]: It's still not clear to me, and someone only passably familiar with the tensor code, how exactly I would use it. Could we have a runnable example showing it end to end with a tensor dataset?

Basically whatever's in the unit test, but in a more friendly example form.

Comment (clarkzinzow) [reply doc/source/data/dataset-tensor-support.rst:249]: Agreed on having an e2e example for tensor data being a good idea, and I'm about to work on creating an e2e example for tensor data as part of the Datasets GA doc work in a separate PR, but the point that I'm making here is that **the user shouldn't have to do anything differently in the `.to_torch()` or `.to_tf()` call just because their data contains tensor columns**.

The tensor column --> ML framework tensor conversion will happen automatically, without any tensor-column-specific args to `.to_torch()` or `.to_tf()`, or special considerations for the upstream tensor column creation other than what's already documented in the tensor column guide: creating the tensor column in the first place. For `.to_tf()`, the tensor spec of the columns needs to be given, but that is already a requirement for all columns.

Since this example work is already planned for before GA, could we merge this in order to unblock the user PoC that depends on this?

Comment (ericl) [reply doc/source/data/dataset-tensor-support.rst:249]: I don't think the example I'm asking for is more in length than this thread already. Let's set a high standard for documentation.

Comment (clarkzinzow) [reply doc/source/data/dataset-tensor-support.rst:249]: I pushed up an example so we can resolve this. I'll have to rewrite/delete this example early next week once other work is merged, which is one of the reasons that I didn't want to block the user PoC on this example, so I'd like it if we could be a bit more pragmatic in the future.

Comment (ericl) [reply doc/source/data/dataset-tensor-support.rst:249]: Thanks. I don't see any harm in blocking this PR on the e2e example, in fact. We should be treating docs like unit tests, and not a separate "optional" component. There shouldn't really be a separate e2e example effort, for example.

I left a separate comment on making the example here more fleshed out (maybe this is copying work from your other effort).

File: doc/source/data/dataset-tensor-support.rst [hunk 7]
```diff
@@ -143,6 +145,77 @@ This dataset can then be written to Parquet files. The tensor column schema will
...
+:meth:`ds.to_tf() <ray.data.Dataset.to_tf>` APIs. The tensor column will be
+automatically converted to a Torch/TensorFlow tensor without incurring any copies.
+
+.. note::
+
+  When converting to a TensorFlow Dataset, you will need to give the full tensor spec
+  for the tensor columns, including the shape of each underlying tensor element in said
+  column.
+
+.. code-block:: python
```
Comment (ericl) [inline doc/source/data/dataset-tensor-support.rst:162]: This isn't really e2e--- could we start with a example dataset generated via `ray.data.range_tensor()` (maybe using add_column to add extra cols), and follow with DataLoader to read the data after?

**Review by ericl — APPROVED** [review]
Approved but I'd prefer we always ship with e2e examples.

**Comment by richardliaw** [comment]
BTW, just to add some color about end-to-end examples.

I think there's a bit of a miscommunication; let me try to elaborate on what @ericl is trying to say.

Clark, from what I understand, is that you are going to build an E2E example very soon, and Eric is seemingly asking for redundant work. 

I think what Eric is asking about is more of a runnable example with more context in the docstrings. 

From the user perspective, I think there's two things that is hard to understand about the current snippets on the tensor user guide:

1. If I create a TF dataset from a tensor dataset, what can I do with it? As a user, what type of inputs will go in, and what type of outputs will I see now? 
2. More broadly, very often, users are going to access docs in random fashion. The current code snippet can be largely improved by just adding a bit of context. Otherwise, as a user, it's the most frustrating thing when you look at docs and you find some random esoteric minimal snippet, and you have no idea what’s it supposed to do and need to piece together a bunch of parts.

(This is a bit of a departure from the current way of writing docs, I know. But I think we can start upping the standards, especially with our new focus on building out great docs. Also, these esoteric minimal snippets are a recent major pain point i've had with Serve and am a bit scarred from that experience ;) )

**Comment by clarkzinzow** [comment]
@ericl @richardliaw I completely agree that this should be the desired end-state! What I'm pushing back against is expanding the scope of this narrow PR to include docs work that is already in progress in other branches/PRs and has already been separately allocated.

To be clear, **the data team is already working on each of the items y'all have mentioned, just not in this PR**.

1. I'm already working on an e2e tensor example (notebook) in another branch/PR.
2. Most of this tensor data feature guide is going to be rewritten and/or split into the other feature guides (e.g. specifics of reading tensor data is probably going to go into the "Creating Datasets" feature guide), which is why I'm not wanting to modify the tensor data feature guide in this PR.
3. The API docs of `to_tf()` and `to_torch()` are going to be overhauled with small e2e examples.
4. We're creating an ML framework integration feature guide that will hold e2e examples and any quirks around handling tensor columns (if there are still any at that point, some WIP work should eliminate some of those existing quirks).

**Comment by richardliaw** [comment]
> What I'm pushing back against is expanding the scope of this narrow PR to include docs work that is already in progress in other branches/PRs and has already been separately allocated.

I think the confusion lies in exactly how much expanded scope we're talking about here (the word "end-to-end" may have a conflated connotations). You'll make me happy with just a small 5-10 line addendum (which I tried to do below, btw, but i really couldn't get it to work):

```diff
diff --git a/doc/source/data/dataset-tensor-support.rst b/doc/source/data/dataset-tensor-support.rst
index 6c6baea07..a268b4cd6 100644
--- a/doc/source/data/dataset-tensor-support.rst
+++ b/doc/source/data/dataset-tensor-support.rst
@@ -161,12 +161,22 @@ automatically converted to a Torch/TensorFlow tensor without incurring any copie

 .. code-block:: python

+  import ray
+  import torch
+
+  ds = ray.data.range_tensor(10000, shape=(3, 5))
+
   # Convert the dataset to a Torch IterableDataset.
   torch_ds = ds.to_torch(
       label_column="label",
       batch_size=4,
       unsqueeze_label_tensor=False,
   )
+  dataloader = torch.utils.data.DataLoader(torch_ds, num_workers=0)
+
+  for x, y in dataloader:
+    # train model(x, y)
+

   # Convert the dataset to a TensorFlow Dataset.
   tensor_element_shape = (32, 32)
```

**Comment by richardliaw** [comment]
The logic from my perspective is the following.

A feature PR should be self-contained with the following things:

1. Correct feature code
2. Tests that verify and maintain the correctness of the functionality of the feature code.
3. Examples or docs or docstrings that allow people to use the feature successfully.

Obviously there's different degrees of investment for 1, 2, and 3. 
Like it'd be crazy to ask for a release test in every feature PR.
And it'd be crazy to ask for a comprehensive e2e example for every feature PR.

But, on the other hand, you still need comprehensive unit tests to verify correctness (which you've done a great job of).
and, you need basic documentation to allow the proficient (or basic) user to be successful (<-- this is the point of contention). 

This PR does not provide enough documentation for even the proficient user to use it successfully.

**Comment by clarkzinzow** [comment]
@richardliaw Again, agreed on those standards for PRs; the data team is in the middle of a massive docs push in order to create complete docs coverage at that standard, and we'll be strictly adhering to that standard post-GA.

In order to unblock this PR and resolve this discussion, I'll add an e2e example to the tensor feature guide.

But in case the motivation I gave above was not clear, I was trying to avoid adding this e2e example to the tensor feature guide since this will end up being redundant work that will be obviated in the next day or two:
1. Today/tomorrow, I'm going to push up code changes to our tensor data handling and `to_torch()`/`to_tf()` that will eliminate the need for an example specifically for tensor data; all tensor data quirks will ideally be removed.
2. Today/tomorrow, I'm going to push up an ML framework exchange feature guide that will house the e2e example and any quirks around data type handling (e.g. tensor data).

Either of the parallel work in (1) or (2) will make this e2e example in the tensor data feature guide obsolete, and I was hoping to quickly get this bug fix to a user doing a PoC, so I was hoping that we could be pragmatic here.

Anyways, I'll add an e2e example to get this unblocked.

**Comment by clarkzinzow** [comment]
> > What I'm pushing back against is expanding the scope of this narrow PR to include docs work that is already in progress in other branches/PRs and has already been separately allocated.
> 
> I think the confusion lies in exactly how much expanded scope we're talking about here (the word "end-to-end" may have a conflated connotations). You'll make me happy with just a small 5-10 line addendum (which I tried to do below, btw, but i really couldn't get it to work):
> 
> ```diff
> diff --git a/doc/source/data/dataset-tensor-support.rst b/doc/source/data/dataset-tensor-support.rst
> index 6c6baea07..a268b4cd6 100644
> --- a/doc/source/data/dataset-tensor-support.rst
> +++ b/doc/source/data/dataset-tensor-support.rst
> @@ -161,12 +161,22 @@ automatically converted to a Torch/TensorFlow tensor without incurring any copie
> 
>  .. code-block:: python
> 
> +  import ray
> +  import torch
> +
> +  ds = ray.data.range_tensor(10000, shape=(3, 5))
> +
>    # Convert the dataset to a Torch IterableDataset.
>    torch_ds = ds.to_torch(
>        label_column="label",
>        batch_size=4,
>        unsqueeze_label_tensor=False,
>    )
> +  dataloader = torch.utils.data.DataLoader(torch_ds, num_workers=0)
> +
> +  for x, y in dataloader:
> +    # train model(x, y)
> +
> 
>    # Convert the dataset to a TensorFlow Dataset.
>    tensor_element_shape = (32, 32)
> ```

@richardliaw As for your e2e example attempt, did you check out the `to_torch()` [docstring](https://docs.ray.io/en/master/data/package-ref.html#ray.data.Dataset.to_torch)? You're not supposed to use the returned `IterableDataset` with a Torch `DataLoader`, you're supposed to consume it directly.

> It is recommended to use the returned IterableDataset directly instead of passing it into a torch DataLoader.

You're also generating a single `"value"` column of tensor data with `ray.data.range_tensor()`, but then specifying a `"label"` label column and expecting there to also be feature columns.

I'm going to be reworking the `to_torch()` and `to_tf()` docstrings in a bit, so let me know if we need to make the `IterableDataset` use (sans `DataLoader`) much more clear, and whether the "only specify a label column, implicitly treat all other columns as feature columns" semantics could be clearer.

I'm guessing that an e2e example in the docstring and the feature guide will make this all much more clear, which is why we have a dedicated docs task for adding these! 😄

**Comment by clarkzinzow** [comment]
@ericl @richardliaw @amogkam Any final feedback on the example or otherwise? If not, I'm planning on merging this today.
